### PR TITLE
refactor(x/gamm, x/swaprouter): multihop and estimate swap move

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This PR moves multihop and estimate swaps keeper methods from x/gamm to x/swaprouter.

No messages or queries are broken.

Old messages and queries relying on the x/gamm keeper methods now directly call the respective methods in `x/swaprouter`.

## Testing and Verifying

All pre-existing tests have been moved to `x/swaprouter`. New tests added where applicable.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -269,6 +269,17 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
...
+		appKeepers.keys[swaproutertypes.StoreKey],
+		appKeepers.GetSubspace(swaproutertypes.ModuleName),
+		appKeepers.GAMMKeeper,
+		nil,
+		appKeepers.BankKeeper,
+		appKeepers.AccountKeeper,
+		appKeepers.DistrKeeper,
+	)
+	appKeepers.GAMMKeeper.SetPoolCreationManager(appKeepers.SwapRouterKeeper)
+
```
Comment (p0mvn) [inline app/keepers/keepers.go:282]:
```suggestion
```

File: x/gamm/keeper/swap.go [hunk 2]
```diff
@@ -49,6 +30,9 @@ func (k Keeper) swapExactAmountIn(
 	if tokenIn.Denom == tokenOutDenom {
 		return sdk.Int{}, errors.New("cannot trade same denomination in and out")
 	}
+	if !pool.IsActive(ctx) {
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:33]: TODO: consider moving this to swaprotuer

File: x/gamm/keeper/grpc_query.go [hunk 3]
```diff
@@ -422,7 +422,10 @@ func (q Querier) EstimateSwapExactAmountIn(ctx context.Context, req *types.Query
 
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
 
-	tokenOutAmount, err := q.Keeper.MultihopEstimateOutGivenExactAmountIn(sdkCtx, req.Routes, tokenIn)
+	// TODO: remove this redundancy after making routes be shared between x/gamm and x/swaprouter.
+	swaprouterRoutes := types.ConvertAmountInRoutes(req.Routes)
```
Comment (p0mvn) [inline x/gamm/keeper/grpc_query.go:426]: Note: this implies proto-generating routes only in swaprouter and importing them in x/gamm.

According to earlier discussions, this should not cause breakage and will be done in a future PR

File: x/swaprouter/router.go [hunk 4]
```diff
@@ -15,15 +18,131 @@ func (k Keeper) RouteExactAmountIn(
 	routes []types.SwapAmountInRoute,
 	tokenIn sdk.Coin,
 	tokenOutMinAmount sdk.Int) (tokenOutAmount sdk.Int, err error) {
-	panic("not implemented")
+	var (
```
Comment (p0mvn) [inline x/swaprouter/router.go:21]: Note: confirmed that this is copied with minimal changes.

The only additional change being choosing the module to route to

File: x/swaprouter/router.go [hunk 5]
```diff
@@ -15,15 +18,131 @@ func (k Keeper) RouteExactAmountIn(
...
+	}
+	return tokenOutAmount, nil
 }
 
 func (k Keeper) MultihopEstimateOutGivenExactAmountIn(
 	ctx sdk.Context,
 	routes []types.SwapAmountInRoute,
 	tokenIn sdk.Coin,
 ) (tokenOutAmount sdk.Int, err error) {
-	panic("not implemented")
```
Comment (p0mvn) [inline x/swaprouter/router.go:26]: Note: confirmed that this is copied with minimal changes.

The only additional changes being:
-  choosing the module to route to
- avoid using sdkerrors

File: x/gamm/keeper/multihop.go [hunk 6]
```diff
@@ -1,365 +0,0 @@
-package keeper
```
Comment (p0mvn) [inline x/gamm/keeper/multihop.go:1]: Note: diffed this whole file against `x/swaprouter/swap.go`. The few changes are summarized in `x/swaprouter/swap.go` per-method.

File: x/swaprouter/router_test.go [hunk 7]
```diff
@@ -25,6 +26,554 @@ var (
...
+				suite.Require().NoError(err)
+				suite.Require().Equal(expectedMultihopTokenOutAmount.Amount.String(), multihopTokenOutAmount.String())
+			}
+		})
+	}
+}
+
+// TestGetSwapModule tests that the correct swap module is returned for a given pool id.
+// Additionally, validates that the expected errors are produced when expected.
+func (suite *KeeperTestSuite) TestGetSwapModule() {
```
Comment (p0mvn) [inline x/swaprouter/router_test.go:512]: TODO: move out into a separate PR

File: x/gamm/keeper/multihop_test.go [hunk 8]
```diff
@@ -1,846 +0,0 @@
-package keeper_test
```
Comment (p0mvn) [inline x/gamm/keeper/multihop_test.go:1]: Note: confirmed that every test case removed in this file has been ported to `x/swaprouter/router_test.go`.

The test fixtures might look different due to clean up. However, every test case is still present

File: x/gamm/keeper/swap.go [hunk 9]
```diff
@@ -151,12 +131,40 @@ func (k Keeper) swapExactAmountOut(
 	return tokenInAmount, nil
 }
 
+func (k Keeper) CalcOutAmtGivenIn(
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:134]: TODO: godoc

File: x/gamm/keeper/swap.go [hunk 10]
```diff
@@ -151,12 +131,40 @@ func (k Keeper) swapExactAmountOut(
...
+	swapFee sdk.Dec,
+) (tokenOut sdk.Coin, err error) {
+	cfmmPool, err := convertToCFMMPool(poolI)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	return cfmmPool.CalcOutAmtGivenIn(ctx, sdk.NewCoins(tokenIn), tokenOutDenom, swapFee)
+}
+
+func (k Keeper) CalcInAmtGivenOut(
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:148]: TODO: godoc

File: x/gamm/types/route.go [hunk 11]
```diff
@@ -91,3 +93,31 @@ type MultihopRoute interface {
 	PoolIds() []uint64
 	IntermediateDenoms() []string
 }
+
+// ConvertAmountInRoutes converts gamm swap exact amount in routes to swaprouter routes.
+// This is a temporary function to be used until we make the route protos be shared between
+// x/gamm and x/swaprouter instead of duplicating them in each module.
+func ConvertAmountInRoutes(gammRoutes []SwapAmountInRoute) []swaproutertypes.SwapAmountInRoute {
```
Comment (p0mvn) [inline x/gamm/types/route.go:100]: TODO: make an issue to track the removal of this

File: x/swaprouter/keeper_test.go [hunk 12]
```diff
@@ -45,12 +42,44 @@ func (suite *KeeperTestSuite) createPoolFromType(poolType types.PoolType) {
...
 		suite.FundAcc(suite.TestAccs[0], curPoolCoins)
-		suite.PrepareBalancerPoolWithCoins(curPoolCoins...)
+		suite.PrepareCustomBalancerPoolFromCoins(curPoolCoins, balancer.PoolParams{
+			SwapFee: swapFee[i],
+			ExitFee: sdk.ZeroDec(),
+		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestInitGenesis() {
```
Comment (p0mvn) [inline x/swaprouter/keeper_test.go:58]: TODO: move into separate PR

File: x/gamm/keeper/swap.go (hunk 2 above)
Comment (p0mvn) [reply x/gamm/keeper/swap.go:33]: https://github.com/osmosis-labs/osmosis/issues/3798

File: x/gamm/types/route.go (hunk 11 above)
Comment (p0mvn) [reply x/gamm/types/route.go:100]: https://github.com/osmosis-labs/osmosis/issues/3799

File: x/swaprouter/types/routes.go [hunk 13]
```diff
@@ -31,43 +31,45 @@ type CommunityPoolI interface {
...
-	// ) (tokenInAmount sdk.Int, err error)
-
-	// CalcInAmtGivenOut(
-	// 	ctx sdk.Context,
-	// 	poolI PoolI,
-	// 	tokenOut sdk.Coin,
-	// 	tokenInDenom string,
-	// 	swapFee sdk.Dec,
-	// ) (tokenIn sdk.Coin, err error)
+	GetPool(ctx sdk.Context, poolId uint64) (PoolI, error)
```
Comment (p0mvn) [inline x/swaprouter/types/routes.go:34]: Note: adding godocs to the rest of these is tracked via: #3556

File: x/gamm/keeper/keeper.go [hunk 14]
```diff
@@ -77,8 +77,8 @@ func (k *Keeper) SetPoolIncentivesKeeper(poolIncentivesKeeper types.PoolIncentiv
 	k.poolIncentivesKeeper = poolIncentivesKeeper
 }
 
-func (k *Keeper) SetPoolCreationManager(poolCreationManager types.PoolCreationManager) {
-	k.poolCreationManager = poolCreationManager
+func (k *Keeper) SetPoolCreationManager(poolCreationManager types.PoolManager) {
+	k.poolManager = poolCreationManager
```
Comment (czarcas7ic) [inline x/gamm/keeper/keeper.go:81]: Can we either stick to the new poolManager nomenclature, or go back to the old poolCreationManager nomenclature?

File: x/gamm/keeper/swap.go [hunk 15]
```diff
@@ -84,30 +73,14 @@ func (k Keeper) swapExactAmountIn(
...
-	pool, err := k.getPoolForSwap(ctx, poolId)
-	if err != nil {
-		return sdk.Int{}, err
-	}
-	swapFee := pool.GetSwapFee(ctx)
-	return k.swapExactAmountOut(ctx, sender, pool, tokenInDenom, tokenInMaxAmount, tokenOut, swapFee)
-}
-
-// swapExactAmountOut is an internal method for swapping to get an exact number of tokens out of a pool,
+// swapExactAmountIn is a method for swapping to get an exact number of tokens out of a pool,
```
Comment (czarcas7ic) [inline x/gamm/keeper/swap.go:76]:
```suggestion
// SwapExactAmountOut is a method for swapping to get an exact number of tokens out of a pool,
```

File: x/gamm/keeper/swap_test.go [hunk 16]
```diff
@@ -114,13 +118,145 @@ func (suite *KeeperTestSuite) TestBalancerPoolSimpleSwapExactAmountIn() {
...
+			if test.expectPass {
+				suite.Require().NoError(err)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+// TestCalcInAmtGivenOut only tests that balancer and stableswap pools are type casted correctly while concentratedliquidity pools fail
```
Comment (czarcas7ic) [inline x/gamm/keeper/swap_test.go:194]: Can we either remove the concentratedliquidity portion from the description, or add a todo to add this to this test as well as the previous one?

File: x/txfees/keeper/keeper.go [hunk 17]
```diff
@@ -16,7 +16,7 @@ type Keeper struct {
 
 	accountKeeper       types.AccountKeeper
 	bankKeeper          types.BankKeeper
-	gammKeeper          types.GammKeeper
+	gammKeeper          types.SwapRouterKeeper
```
Comment (czarcas7ic) [inline x/txfees/keeper/keeper.go:19]: is the plan to keep this as gammKeeper utilizing swaprouterkeeper, or does this eventually change?

**Review by czarcas7ic — APPROVED** [review]
Well done, only minor comments on my end 👍

File: x/gamm/keeper/swap_test.go (hunk 16 above)
Comment (p0mvn) [reply x/gamm/keeper/swap_test.go:195]: added TODO, including here: https://github.com/osmosis-labs/osmosis/issues/3556

File: x/txfees/types/expected_keepers.go [hunk 18]
```diff
@@ -11,16 +13,14 @@ type SpotPriceCalculator interface {
...
+	RouteExactAmountIn(
 		ctx sdk.Context,
 		sender sdk.AccAddress,
-		poolId uint64,
+		routes []swaproutertypes.SwapAmountInRoute,
 		tokenIn sdk.Coin,
-		tokenOutDenom string,
-		tokenOutMinAmount sdk.Int,
-	) (tokenOutAmount sdk.Int, err error)
+		tokenOutMinAmount sdk.Int) (tokenOutAmount sdk.Int, err error)
```
Comment (ValarDragon) [inline x/txfees/types/expected_keepers.go:23]: Can we keep an API shim for single hops

Comment (ValarDragon) [reply x/txfees/types/expected_keepers.go:23]: (can be done in a subsequent PR)

Comment (ValarDragon) [reply x/txfees/types/expected_keepers.go:23]: Also why was this renamed to RouteExactAmountIn?

Comment (p0mvn) [reply x/txfees/types/expected_keepers.go:23]: Sounds good, will make an issue to create a separate API.

In terms of abstraction, what this method does is routes swaps across pools over several modules so the rename made sense.

I don't have a strong opinion and can change this back to multihop if you prefer.

File: x/gamm/keeper/grpc_query.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/keeper/grpc_query.go:426]: SG if were confident it wont break stuff. Is there an issue or something written to track our testing strategy? (Not PR merge blocking)

File: x/gamm/keeper/pool_service.go [hunk 19]
```diff
@@ -342,12 +342,19 @@ func (k Keeper) ExitSwapShareAmountIn(
...
+	}
+	swapFee := pool.GetSwapFee(ctx)
+
 	tokenOutAmount = exitCoins.AmountOf(tokenOutDenom)
 	for _, coin := range exitCoins {
 		if coin.Denom == tokenOutDenom {
 			continue
 		}
-		swapOut, err := k.SwapExactAmountIn(ctx, sender, poolId, coin, tokenOutDenom, sdk.ZeroInt())
+		swapOut, err := k.SwapExactAmountIn(ctx, sender, pool, coin, tokenOutDenom, sdk.ZeroInt(), swapFee)
```
Comment (ValarDragon) [inline x/gamm/keeper/pool_service.go:357]: Wait, theres been an important abstraction issue thats changed here wrt swap fee.

I get why its being done here, but am concerned about:
- Whats the public/private guarantee
- Ensuring we keep dynamic swap fees possible.

The latter is still the case, if AMMs can "ignore" the swapFee arg.

Maybe we add a panic for now, if swapFee arg less than .5 ctx.SwapFee() ?

I think its good to merge if we add that safety check, and add to our mental checklist, that modules shouldn't import gamm, only swaprouter.

File: x/gamm/types/expected_keepers.go [hunk 20]
```diff
@@ -44,11 +44,39 @@ type CommunityPoolKeeper interface {
...
+
+	MultihopEstimateInGivenExactAmountOut(
+		ctx sdk.Context,
+		routes []swaproutertypes.SwapAmountOutRoute,
+		tokenOut sdk.Coin) (tokenInAmount sdk.Int, err error)
+}
+
+type PoolIncentivesKeeper interface {
+	IsPoolIncentivized(ctx sdk.Context, poolId uint64) bool
 }
```
Comment (ValarDragon) [inline x/gamm/types/expected_keepers.go:82]: This should've moved to swaprouter, no?

**Review by ValarDragon — APPROVED** [review]
LGTM, aside from my comments.

I am assuming the router.go and router_test.go are pure moves, but I have not in-depth verified this.

File: x/txfees/types/expected_keepers.go (hunk 18 above)
Comment (p0mvn) [reply x/txfees/types/expected_keepers.go:23]: https://github.com/osmosis-labs/osmosis/issues/3807

File: x/gamm/keeper/pool_service.go (hunk 19 above)
Comment (p0mvn) [reply x/gamm/keeper/pool_service.go:357]: I have the same concern and this is tracked here: https://github.com/osmosis-labs/osmosis/issues/3130

I've added the suggested check and tests for it in: [ace13f9](https://github.com/osmosis-labs/osmosis/pull/3764/commits/ace13f91ff1f2f79dffe0966fc2579ccd3af33e1)

File: x/gamm/keeper/swap.go [hunk 21]
```diff
@@ -130,7 +105,12 @@ func (k Keeper) swapExactAmountOut(
 			"can't get more tokens out than there are tokens in the pool")
 	}
 
-	tokenIn, err := pool.SwapInAmtGivenOut(ctx, sdk.Coins{tokenOut}, tokenInDenom, swapFee)
+	cfmmPool, err := convertToCFMMPool(pool)
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/gamm/keeper/swap.go:112]: This type cast happens frequently so made sense to abstract it away

File: x/gamm/keeper/grpc_query.go (hunk 3 above)
Comment (p0mvn) [reply x/gamm/keeper/grpc_query.go:426]: According to our earlier discussions, this should not be state-breaking. 

The only concern was breaking telescope. I'm planning to make a follow-up PR and tag Dan Lynch to help with looking into this.

I can also add a test to serialize both old and new routes to make sure that they are valid

Comment (p0mvn) [reply x/gamm/keeper/grpc_query.go:426]: https://github.com/osmosis-labs/osmosis/issues/3808
